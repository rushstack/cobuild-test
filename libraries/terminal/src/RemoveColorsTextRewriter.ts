// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { AnsiEscape } from '@rushstack/node-core-library';
import { TextRewriter, TextRewriterState } from './TextRewriter';

enum State {
  // Buffer is empty, and we're looking for the ESC character
  Start,
  // We're looking for the '[' character
  AwaitingBracket,
  // We're reading the codes after the '[' character
  ReadingCodes
}

interface IRemoveColorsTextRewriterState extends TextRewriterState {
  buffer: string;
  parseState: State;
}

/**
 * For use with {@link TextRewriterTransform}, this rewriter removes ANSI escape codes
 * including colored text.
 *
 * @remarks
 * The implementation also removes other ANSI escape codes such as cursor positioning.
 * The specific set of affected codes may be adjusted in the future.
 *
 * @public
 */
export class RemoveColorsTextRewriter extends TextRewriter {
  public initialize(): TextRewriterState {
    return { buffer: '', parseState: State.Start } as IRemoveColorsTextRewriterState;
  }

  public process(unknownState: TextRewriterState, text: string): string {
    const state: IRemoveColorsTextRewriterState = unknownState as IRemoveColorsTextRewriterState;

    // We will be matching AnsiEscape._csiRegExp:
    //
    //  /\x1b\[([\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])/gu
    //
    const ESC: string = '\x1b';

    let result: string = '';
    let index: number = 0;

    while (index < text.length) {
      if (state.parseState === State.Start) {
        // The buffer is empty, which means we haven't found anything yet

        const csiIndex: number = text.indexOf(ESC, index);
        if (csiIndex < 0) {
          // We reached the end of "text" without finding another CSI prefix
          result += text.substring(index);
          break;
        }

        // Append everything up to the CSI prefix
        result += text.substring(index, csiIndex);

        // Save the partial match in the buffer
        state.buffer = ESC;
        index = csiIndex + 1;
        state.parseState = State.AwaitingBracket;
      } else {
        // The buffer has characters, which means we started matching a partial sequence

        // Read another character into the buffer
        const c: string = text[index];
        ++index;
        state.buffer += c;

        if (state.parseState === State.AwaitingBracket) {
          if (c === '[') {
            state.parseState = State.ReadingCodes;
          } else {
            // Failed to match, so append the buffer and start over
            result += state.buffer;
            state.buffer = '';
            state.parseState = State.Start;
          }
        } else {
          // state.state === State.ReadingCodes

          // Stop when we reach any character that is not [\x30-\x3f] or [\x20-\x2f]
          const code: number = c.charCodeAt(0);
          if (code < 0x20 || code > 0x3f) {
            result += AnsiEscape.removeCodes(state.buffer);
            state.buffer = '';
            state.parseState = State.Start;
          }
        }
      }
    }

    return result;
  }

  public close(unknownState: TextRewriterState): string {
    const state: IRemoveColorsTextRewriterState = unknownState as IRemoveColorsTextRewriterState;

    const result: string = state.buffer;
    state.buffer = '';
    return result;
  }
}
