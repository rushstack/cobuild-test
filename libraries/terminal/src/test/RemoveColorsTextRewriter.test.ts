// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import colors from 'colors';

import { RemoveColorsTextRewriter } from '../RemoveColorsTextRewriter';
import { TextRewriterState } from '../TextRewriter';
import { AnsiEscape } from '@rushstack/node-core-library';

function testCase(inputs: string[]): void {
  const matcher: RemoveColorsTextRewriter = new RemoveColorsTextRewriter();
  const state: TextRewriterState = matcher.initialize();
  const outputs: string[] = inputs.map((x) => matcher.process(state, x));
  const closeOutput: string = matcher.close(state);
  if (closeOutput !== '') {
    outputs.push('--close--');
    outputs.push(closeOutput);
  }

  expect({
    inputs: inputs.map((x) => AnsiEscape.formatForTests(x)),
    outputs
  }).toMatchSnapshot();
}

describe(RemoveColorsTextRewriter.name, () => {
  let initialColorsEnabled: boolean;

  beforeAll(() => {
    initialColorsEnabled = colors.enabled;
    colors.enable();
  });

  afterAll(() => {
    if (!initialColorsEnabled) {
      colors.disable();
    }
  });

  it('01 should process empty inputs', () => {
    testCase([]);
    testCase(['']);
    testCase(['', 'a', '']);
  });

  it('02 should remove colors from complete chunks', () => {
    testCase([colors.red('1')]);
    testCase([colors.red('1') + colors.green('2')]);
    testCase([colors.red('1') + '2' + colors.green('3')]);
  });

  it('03 should remove colors from 1-character chunks', () => {
    const source: string = '1' + colors.red('2');
    const inputs: string[] = [];
    for (let i: number = 0; i < source.length; ++i) {
      inputs.push(source.substr(i, 1));
    }
    testCase(inputs);
  });

  it('04 should pass through incomplete partial matches', () => {
    testCase(['\x1b']);
    testCase(['\x1b[\n']);
    testCase(['\x1b[1']);
  });
});
