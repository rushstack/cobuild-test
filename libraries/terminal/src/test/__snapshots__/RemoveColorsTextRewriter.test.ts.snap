// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveColorsTextRewriter 01 should process empty inputs 1`] = `
Object {
  "inputs": Array [],
  "outputs": Array [],
}
`;

exports[`RemoveColorsTextRewriter 01 should process empty inputs 2`] = `
Object {
  "inputs": Array [
    "",
  ],
  "outputs": Array [
    "",
  ],
}
`;

exports[`RemoveColorsTextRewriter 01 should process empty inputs 3`] = `
Object {
  "inputs": Array [
    "",
    "a",
    "",
  ],
  "outputs": Array [
    "",
    "a",
    "",
  ],
}
`;

exports[`RemoveColorsTextRewriter 02 should remove colors from complete chunks 1`] = `
Object {
  "inputs": Array [
    "[red]1[default]",
  ],
  "outputs": Array [
    "1",
  ],
}
`;

exports[`RemoveColorsTextRewriter 02 should remove colors from complete chunks 2`] = `
Object {
  "inputs": Array [
    "[red]1[default][green]2[default]",
  ],
  "outputs": Array [
    "12",
  ],
}
`;

exports[`RemoveColorsTextRewriter 02 should remove colors from complete chunks 3`] = `
Object {
  "inputs": Array [
    "[red]1[default]2[green]3[default]",
  ],
  "outputs": Array [
    "123",
  ],
}
`;

exports[`RemoveColorsTextRewriter 03 should remove colors from 1-character chunks 1`] = `
Object {
  "inputs": Array [
    "1",
    "",
    "[",
    "3",
    "1",
    "m",
    "2",
    "",
    "[",
    "3",
    "9",
    "m",
  ],
  "outputs": Array [
    "1",
    "",
    "",
    "",
    "",
    "",
    "2",
    "",
    "",
    "",
    "",
    "",
  ],
}
`;

exports[`RemoveColorsTextRewriter 04 should pass through incomplete partial matches 1`] = `
Object {
  "inputs": Array [
    "",
  ],
  "outputs": Array [
    "",
    "--close--",
    "",
  ],
}
`;

exports[`RemoveColorsTextRewriter 04 should pass through incomplete partial matches 2`] = `
Object {
  "inputs": Array [
    "[
",
  ],
  "outputs": Array [
    "[
",
  ],
}
`;

exports[`RemoveColorsTextRewriter 04 should pass through incomplete partial matches 3`] = `
Object {
  "inputs": Array [
    "[1",
  ],
  "outputs": Array [
    "",
    "--close--",
    "[1",
  ],
}
`;
